flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/PostTransformOperator.java

```diff
@@ -434,23 +436,36 @@ private Optional<DataChangeEvent> processDataChangeEvent(DataChangeEvent dataCha
                         && transformProjectionOptional.get().isValid()) {
                     TransformProjection transformProjection = transformProjectionOptional.get();
                     if (!transformProjectionProcessorMap.containsKey(
-                                    Tuple2.of(tableId, transformProjection))
-                            || !transformProjectionProcessorMap
-                                    .get(Tuple2.of(tableId, transformProjection))
-                                    .hasTableInfo()) {
+                            Tuple2.of(tableId, transformProjection.getProjection()))) {
                         transformProjectionProcessorMap.put(
-                                Tuple2.of(tableId, transformProjection),
+                                Tuple2.of(tableId, transformProjection.getProjection()),
                                 TransformProjectionProcessor.of(
                                         tableInfo,
                                         transformProjection,
                                         timezone,
                                         udfDescriptors,
                                         getUdfFunctionInstances(),
                                         transform.getSupportedMetadataColumns()));
+                    } else if (!transformProjectionProcessorMap
+                            .get(Tuple2.of(tableId, transformProjection.getProjection()))
+                            .hasTableInfo()) {
```

Comment:
`hasTableInfo` could be renamed to something like `hasChangeInfo` or `hasPostTransformChangeInfo`